File: frontend/src/app/services/guia-genetica.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, of } from 'rxjs';
import { map, catchError } from 'rxjs/operators';
import { environment } from '../../environments/environment';

export interface GuiaGeneticaDto {
  edad: number;
  consumoHembras: number;      // Gramos por ave por día
  consumoMachos: number;       // Gramos por ave por día
  pesoHembras: number;         // Peso esperado hembras
  pesoMachos: number;          // Peso esperado machos
  mortalidadHembras: number;   // Mortalidad esperada hembras
  mortalidadMachos: number;    // Mortalidad esperada machos
  uniformidad: number;         // Uniformidad esperada
  pisoTermicoRequerido: boolean; // Si requiere piso térmico
  observaciones?: string;      // Observaciones adicionales
}

export interface GuiaGeneticaRequest {
  raza: string;
  anoTabla: number;
  edad: number;
}

export interface GuiaGeneticaResponse {
  existe: boolean;
  datos?: GuiaGeneticaDto;
  mensaje?: string;
}

@Injectable({
  providedIn: 'root'
})
export class GuiaGeneticaService {
  private readonly apiUrl = `${environment.apiUrl}/guia-genetica`;

  constructor(private http: HttpClient) {}

  /**
   * Obtiene datos de guía genética para una edad específica
   */
  obtenerGuiaGenetica(raza: string, anoTabla: number, edad: number): Observable<GuiaGeneticaResponse> {
    const params = new HttpParams()
      .set('raza', raza)
      .set('anoTabla', anoTabla.toString())
      .set('edad', edad.toString());

    return this.http.get<GuiaGeneticaResponse>(`${this.apiUrl}/obtener`, { params });
  }

  /**
   * Obtiene datos de guía genética para un rango de edades
   */
  obtenerGuiaGeneticaRango(raza: string, anoTabla: number, edadDesde: number, edadHasta: number): Observable<GuiaGeneticaDto[]> {
    const params = new HttpParams()
      .set('raza', raza)
      .set('anoTabla', anoTabla.toString())
      .set('edadDesde', edadDesde.toString())
      .set('edadHasta', edadHasta.toString());

    return this.http.get<GuiaGeneticaDto[]>(`${this.apiUrl}/rango`, { params });
  }

  /**
   * Verifica si existe una guía genética
   */
  existeGuiaGenetica(raza: string, anoTabla: number): Observable<boolean> {
    const params = new HttpParams()
      .set('raza', raza)
      .set('anoTabla', anoTabla.toString());

    return this.http.get<boolean>(`${this.apiUrl}/existe`, { params });
  }

  /**
   * Obtiene las razas disponibles en las guías genéticas
   * Consume directamente desde la base de datos produccion_avicola_raw
   */
  obtenerRazasDisponibles(): Observable<string[]> {
    return this.http.get<string[]>(`${this.apiUrl}/razas`).pipe(
      map(razas => {
        // Filtrar solo las razas válidas de la guía genética
        const razasValidas = razas.filter(raza => {
          const razaTrimmed = raza?.trim();
          // Solo incluir razas que parecen ser códigos de guía genética válidos
          return razaTrimmed && 
                 razaTrimmed !== '' && 
                 razaTrimmed !== 'AP' && // Excluir datos hardcodeados
                 razaTrimmed !== 'C500' && // Excluir datos hardcodeados
                 razaTrimmed.length >= 2; // Al menos 2 caracteres
        });
        
        return razasValidas
          .map(raza => raza.trim())
          .sort((a, b) => a.localeCompare(b));
      }),
      catchError(error => {
        console.error('Error obteniendo razas disponibles:', error);
        return of([]); // Retornar array vacío en caso de error
      })
    );
  }

  /**
   * Obtiene los años disponibles para una raza específica
   * Consume directamente desde la base de datos produccion_avicola_raw
   */
  obtenerAnosDisponibles(raza: string): Observable<number[]> {
    if (!raza || raza.trim() === '') {
      return of([]);
    }

    const params = new HttpParams().set('raza', raza.trim());
    return this.http.get<number[]>(`${this.apiUrl}/anos`, { params }).pipe(
      map(anos => {
        // Filtrar años válidos y ordenar descendente (más recientes primero)
        return anos
          .filter(ano => ano && ano > 2000 && ano <= new Date().getFullYear() + 1)
          .sort((a, b) => b - a);
      }),
      catchError(error => {
        console.error('Error obteniendo años disponibles:', error);
        return of([]); // Retornar array vacío en caso de error
      })
    );
  }

  /**
   * Valida si una combinación de raza y año es válida
   */
  validarCombinacionRazaAno(raza: string, anoTabla: number): Observable<boolean> {
    if (!raza || raza.trim() === '' || !anoTabla || anoTabla <= 0) {
      return of(false);
    }

    return this.existeGuiaGenetica(raza.trim(), anoTabla);
  }

  /**
   * Obtiene información completa de una raza (años disponibles y validación)
   */
  obtenerInformacionRaza(raza: string): Observable<{
    raza: string;
    anosDisponibles: number[];
    esValida: boolean;
  }> {
    if (!raza || raza.trim() === '') {
      return of({
        raza: '',
        anosDisponibles: [],
        esValida: false
      });
    }

    return this.obtenerAnosDisponibles(raza).pipe(
      map(anos => ({
        raza: raza.trim(),
        anosDisponibles: anos,
        esValida: anos.length > 0
      })),
      catchError(error => {
        console.error('Error obteniendo información de raza:', error);
        return of({
          raza: raza.trim(),
          anosDisponibles: [],
          esValida: false
        });
      })
    );
  }

  /**
   * Método de conveniencia para obtener consumo esperado
   */
  async obtenerConsumoEsperado(raza: string, anoTabla: number, edad: number): Promise<number> {
    try {
      const response = await this.obtenerGuiaGenetica(raza, anoTabla, edad).toPromise();
      
      if (response?.existe && response.datos) {
        // Promedio entre hembras y machos
        return (response.datos.consumoHembras + response.datos.consumoMachos) / 2;
      }
      
      return 0; // Valor por defecto si no existe la guía
    } catch (error) {
      console.error('Error obteniendo consumo esperado:', error);
      return 0;
    }
  }

  /**
   * Método de conveniencia para verificar si requiere piso térmico
   */
  async requierePisoTermico(raza: string, anoTabla: number, edad: number): Promise<boolean> {
    try {
      const response = await this.obtenerGuiaGenetica(raza, anoTabla, edad).toPromise();
      
      if (response?.existe && response.datos) {
        return response.datos.pisoTermicoRequerido;
      }
      
      return edad <= 3; // Valor por defecto: primeras 3 semanas
    } catch (error) {
      console.error('Error verificando piso térmico:', error);
      return edad <= 3;
    }
  }

  /**
   * Método de conveniencia para obtener peso esperado
   */
  async obtenerPesoEsperado(raza: string, anoTabla: number, edad: number): Promise<{ pesoHembras: number; pesoMachos: number }> {
    try {
      const response = await this.obtenerGuiaGenetica(raza, anoTabla, edad).toPromise();
      
      if (response?.existe && response.datos) {
        return {
          pesoHembras: response.datos.pesoHembras,
          pesoMachos: response.datos.pesoMachos
        };
      }
      
      return { pesoHembras: 0, pesoMachos: 0 };
    } catch (error) {
      console.error('Error obteniendo peso esperado:', error);
      return { pesoHembras: 0, pesoMachos: 0 };
    }
  }
}

File: backend/src/ZooSanMarino.Infrastructure/Services/GuiaGeneticaService.cs
using Microsoft.EntityFrameworkCore;
using ZooSanMarino.Application.DTOs;
using ZooSanMarino.Application.Interfaces;
using ZooSanMarino.Domain.Entities;
using ZooSanMarino.Infrastructure.Persistence;

namespace ZooSanMarino.Infrastructure.Services;

public class GuiaGeneticaService : IGuiaGeneticaService
{
    private readonly ZooSanMarinoContext _ctx;

    public GuiaGeneticaService(ZooSanMarinoContext ctx)
    {
        _ctx = ctx;
    }

    /// <summary>
    /// Obtiene los datos de guía genética para una raza, año y edad específicos
    /// </summary>
    public async Task<GuiaGeneticaResponse> ObtenerGuiaGeneticaAsync(GuiaGeneticaRequest request)
    {
        try
        {
            // Buscar en la tabla de guías genéticas
            var guia = await _ctx.ProduccionAvicolaRaw
                .Where(p => p.Raza == request.Raza && 
                           p.AnioGuia == request.AnoTabla.ToString() && 
                           p.Edad == request.Edad.ToString())
                .FirstOrDefaultAsync();

            if (guia == null)
            {
                return new GuiaGeneticaResponse(
                    Existe: false,
                    Datos: null,
                    Mensaje: $"No se encontró guía genética para Raza: {request.Raza}, Año: {request.AnoTabla}, Edad: {request.Edad}"
                );
            }

            // Parsear los valores de la guía genética
            var datos = new GuiaGeneticaDto(
                Edad: request.Edad,
                ConsumoHembras: ParseDouble(guia.GrAveDiaH),
                ConsumoMachos: ParseDouble(guia.GrAveDiaM),
                PesoHembras: ParseDouble(guia.PesoH),
                PesoMachos: ParseDouble(guia.PesoM),
                MortalidadHembras: ParseDouble(guia.MortSemH),
                MortalidadMachos: ParseDouble(guia.MortSemM),
                Uniformidad: ParseDouble(guia.Uniformidad),
                PisoTermicoRequerido: DeterminarPisoTermico(request.Edad, guia),
                Observaciones: $"Guía: {guia.Raza} {guia.AnioGuia}"
            );

            return new GuiaGeneticaResponse(
                Existe: true,
                Datos: datos,
                Mensaje: "Guía genética encontrada exitosamente"
            );
        }
        catch (Exception ex)
        {
            return new GuiaGeneticaResponse(
                Existe: false,
                Datos: null,
                Mensaje: $"Error al obtener guía genética: {ex.Message}"
            );
        }
    }

    /// <summary>
    /// Obtiene múltiples edades de una guía genética
    /// </summary>
    public async Task<IEnumerable<GuiaGeneticaDto>> ObtenerGuiaGeneticaRangoAsync(string raza, int anoTabla, int edadDesde, int edadHasta)
    {
        var guias = await _ctx.ProduccionAvicolaRaw
            .Where(p => p.Raza == raza && 
                       p.AnioGuia == anoTabla.ToString())
            .ToListAsync();

        return guias
            .Where(p => int.TryParse(p.Edad, out var edad) && 
                       edad >= edadDesde && edad <= edadHasta)
            .OrderBy(p => int.Parse(p.Edad ?? "0"))
            .Select(g => new GuiaGeneticaDto(
                Edad: int.Parse(g.Edad ?? "0"),
                ConsumoHembras: ParseDouble(g.GrAveDiaH),
                ConsumoMachos: ParseDouble(g.GrAveDiaM),
                PesoHembras: ParseDouble(g.PesoH),
                PesoMachos: ParseDouble(g.PesoM),
                MortalidadHembras: ParseDouble(g.MortSemH),
                MortalidadMachos: ParseDouble(g.MortSemM),
                Uniformidad: ParseDouble(g.Uniformidad),
                PisoTermicoRequerido: DeterminarPisoTermico(int.Parse(g.Edad ?? "0"), g),
                Observaciones: $"Guía: {g.Raza} {g.AnioGuia}"
            ));
    }

    /// <summary>
    /// Verifica si existe una guía genética para los parámetros dados
    /// </summary>
    public async Task<bool> ExisteGuiaGeneticaAsync(string raza, int anoTabla)
    {
        return await _ctx.ProduccionAvicolaRaw
            .AnyAsync(p => p.Raza == raza && p.AnioGuia == anoTabla.ToString());
    }

    /// <summary>
    /// Obtiene las razas disponibles en las guías genéticas
    /// </summary>
    public async Task<IEnumerable<string>> ObtenerRazasDisponiblesAsync()
    {
        var razas = await _ctx.ProduccionAvicolaRaw
            .Where(p => !string.IsNullOrEmpty(p.Raza))
            .Select(p => p.Raza!)
            .Distinct()
            .ToListAsync();

        // Filtrar solo las razas que parecen ser códigos de guía genética válidos
        // Mantener todas las razas disponibles en la base de datos
        var razasValidas = razas
            .Where(raza => !string.IsNullOrEmpty(raza) && 
                          raza.Trim().Length >= 2)
            .Select(raza => raza.Trim())
            .OrderBy(r => r)
            .ToList();

        return razasValidas;
    }

    /// <summary>
    /// Obtiene los años disponibles para una raza específica
    /// </summary>
    public async Task<IEnumerable<int>> ObtenerAnosDisponiblesAsync(string raza)
    {
        var anos = await _ctx.ProduccionAvicolaRaw
            .Where(p => p.Raza == raza && !string.IsNullOrEmpty(p.AnioGuia))
            .Select(p => p.AnioGuia!)
            .Distinct()
            .ToListAsync();

        // Parsear de forma segura y filtrar solo los años válidos
        var anosValidos = anos
            .Where(ano => int.TryParse(ano, out _))
            .Select(ano => int.Parse(ano))
            .OrderByDescending(a => a)
            .ToList();

        return anosValidos;
    }

    // ================== MÉTODOS PRIVADOS ==================

    /// <summary>
    /// Parsea un string a double de forma segura
    /// </summary>
    private static double ParseDouble(string? value)
    {
        if (string.IsNullOrEmpty(value))
            return 0.0;

        // Remover caracteres no numéricos excepto punto y coma
        var cleanValue = value.Replace(",", ".").Trim();
        
        if (double.TryParse(cleanValue, out var result))
            return result;

        return 0.0;
    }

    /// <summary>
    /// Determina si se requiere piso térmico basado en la edad y datos de la guía
    /// </summary>
    private static bool DeterminarPisoTermico(int edad, ProduccionAvicolaRaw guia)
    {
        // Lógica para determinar piso térmico:
        // 1. Si la edad es menor o igual a 3 semanas
        // 2. O si hay información específica en la guía
        
        if (edad <= 3)
            return true;

        // Verificar si hay información específica en campos adicionales
        // Por ejemplo, si Valor1000 contiene información sobre temperatura
        if (!string.IsNullOrEmpty(guia.Valor1000))
        {
            var valor = guia.Valor1000.ToLower();
            if (valor.Contains("termico") || valor.Contains("calor") || valor.Contains("temperatura"))
                return true;
        }

        return false;
    }
}
